app/javascript/miq-redux/reducer.ts

```diff
@@ -0,0 +1,25 @@
+import {createStore, Store, Reducer, Action} from 'redux';
+
+import { IMiqAction, AppReducer, IMiqReducerHash, AppState} from '../packs/miq-redux';
+
+const reducers: Set<AppReducer> = new Set();
+
+export const rootReducer: AppReducer = (state = {}, action: IMiqAction) => {
+    let newState = state;
+
+    reducers.forEach((reducer) => {
+        newState = reducer(newState, action)
```

Comment:
We basically try to aim for simplicity and flexibility.

The `addReducer` (miq-redux) function is to be used by both core (manageiq-ui-classic) and plugins. You start with no reducers (empty `Set`) and allow client code to add new ones.

In a component-oriented architecture with Redux, you have specific components bound to (or data-driven by) specific parts of the state tree. Having one reducer per one logical component seems like a natural fit for such architecture.

As you wrote, one would normally use Redux `combineReducers` helper to combine smaller reducers into bigger ones. The difference here is that we don't (and shouldn't) know in advance what the specific reducers will be. We simply provide a way to add your reducer via `addReducer`. This means the MiQ Redux infra is opt-in by design, allowing gradual adoption of Redux architecture, typically on per-component basis.

Calling all reducers in a `forEach` loop reflects the design that "all reducers are on the same level" - they all get the whole state, the action that was dispatched, and impose changes on the relevant part of the state tree (while following the general principle of state immutability). This design follows the goals mentioned above - simplicity and flexibility.

We could, by design, constrain each reducer to a specific state subtree. It would probably result in API like `addReducer(myReducer, 'foo.bar')` where `myReducer` would be constrained to operate only on a subtree represented by the expression `state.foo.bar`. But do we really need that?

All reducers, both core and plugin contributed, are to be treated in the same way from MiQ Redux infra perspective. If a reducer does something harmful, like overwriting a specific state subtree in an inconsistent way, it's a problem with that particular reducer (or generally with the code that added such reducer).